### ISSUE TITLE
Type 'StitchClientErrorCode' is not assignable to type 'string'. Angular 6

### ISSUE THREAD

**Issue by saniyusuf** [mongodb/stitch-js-sdk#150]
I have just recommended Stitch to my company. We are testing and it seems the latest stitch client is broken and won't let my Angular test application compile. 
As soon as I put the code that creates a new Stitch client,  the webpack compilation fails. 
`
import { Stitch } from "mongodb-stitch-browser-sdk";

const client = Stitch.initializeDefaultAppClient('strat-care-development-zaxso');
`
This is literally the first line without even providing anything. From the logs, it looks like a dependecy issue with the "detect-browser" package. Below is the error. 

`ERROR in ./node_modules/detect-browser/index.js
`

Below is my Angular Environment info. 
Angular CLI: 6.0.8
Node: 8.11.1
OS: darwin x64
Angular: 6.0.6

**Comment by edaniels** [comment]
Hi! Sorry you hit that. I see your using TS 2.7.2 but I'm curious if @angular/cli is using the same version. Could you verify what version it's using in your package-lock.json? If it's the same we can do some more troubleshooting.

**Comment by edaniels** [comment]
For reference, we use TypeScript 2.9.2

**Comment by saniyusuf** [comment]
@edaniels This is correct. I am using TS 2.7.2 as this is what Angular pulls out. Is this the reason why this issue is happening? In my Package JSON I see version ~2.7.2. In the lock, there is 2.9.2 under "@angular-devkit/build-optimizer". 

Any ideas on how I can solve this problem. I will be honest I am not sure what is going on here.

**Comment by edaniels** [comment]
Can you upgrade your typescript version to 2.9.2? If not, you can try changing your tsconfig compilerOptions to reference --types (https://www.typescriptlang.org/docs/handbook/compiler-options.html) to include the only type definitions you want to include. As far as I can tell, anguular is including Stitch's TypeScript definitions which are not compatible with TS 2.7.2. I'm getting some of this info from https://github.com/Microsoft/TypeScript/issues/9731.

I believe Angular is going to 2.9.2 very soon by the way.

**Comment by saniyusuf** [comment]
@edaniels I have already tried using 2.9.2, the compiler would complain due to other angular dependencies needing that version. I want to try the second option but, I am not sure what types to include. Do you have a list maybe? Im sorry for my ignorant view on the matter as I have limited experience with tooling level stuff.

**Comment by edaniels** [comment]
Well if you set types to `[]` and try again, you will probably start getting notifications as to what packages don't have types and you can start including as you go. It will depend on what your app is specifically using.

**Comment by saniyusuf** [comment]
@edaniels I have tried this fix and the error remains the same. I have a question. Is there a way to use an olde version of the stitch client? 

If there is, I would be happy to do this for now. The links for these are broken at the moment.

**Comment by edaniels** [comment]
Oh yeah sure. We don't support the 3.x series but there's `mongodb-stitch`: `3.1.7`

**Comment by saniyusuf** [comment]
@edaniels is there a link to this? Any link to documentation or anything will help.

**Comment by edaniels** [comment]
Yep right here https://www.npmjs.com/package/mongodb-stitch

**Comment by saniyusuf** [comment]
@edaniels Thanks for your help. I have found a quick fix. The best way to get the client working is by using the JS SDK. I downloaded a copy of the CDN and packaged it as part of my build and with a bit of careful programming since there is no intellisense, I have been able to connect. 

This should allow anyone do everything till Angular moves to TS 2.9.2

**Comment by edaniels** [comment]
Fantastic. Thank you @saniyusuf.

**Comment by crafter999** [comment]
### Quick Dirty Angular 6 Fix
I don't have enough time to evaluate every single line of code and make it work without causing any trouble but following these dirty fixes **will** make Angular 6 (including intellisence) work but i can't guarantee any stable results. Maybe I'm wrong and there's no need for any of those actions but I couldn't find any solution.

First of all, the module detect-browser requires a package named 'os' (I don't know why they didn't included it at their `package.json` but if you install it, the error goes away)
Secondly, you should consider making the enum `StitchClientErrorCode` to something more flexible cause they're throwing the following errors. You can initialize them as strings and the type errors goes away. 

```
ERROR in node_modules/mongodb-stitch-core-sdk/dist/esm/StitchClientErrorCode.d.ts(9,12): error TS2322: Type 'StitchClientErrorCode' is not assignable to type 'string'.
node_modules/mongodb-stitch-core-services-mongodb-remote/dist/esm/RemoteInsertManyResult.d.ts(2,34): error TS2344: Type 'number' does not satisfy the constraint 'string'.
```


Lastly for all the Angular users who get "Buffer is not defined" they should add the following lines to their `polyfills.ts`.

```
declare var global: any;
declare var require: any;

(window as any).global = window;
global.Buffer = global.Buffer || require('buffer').Buffer;
```


I know there fixes are really dirty and need some real evaluation but there's no other way to make it work with Angular 6, I posted a patch here and if you really need a pull request on the actual project (and not these dists files) I could make one.
Patch -> https://gist.github.com/crafter999/b60b2060c785c2bc005aa46eca3821d4

### Recap for Angular 6 users (TLDR)
1: Create a new project
2: Install stitch `npm install -save-dev mongodb-stitch-browser-sdk`
3: Apply the patch from https://gist.github.com/crafter999/b60b2060c785c2bc005aa46eca3821d4
4: Install the os package `npm install --save-dev os`
5: Import stitch and you're ready to go `import * as stitch from 'mongodb-stitch-browser-sdk';`

**Comment by edaniels** [comment]
Thanks @crafter999. We're hoping angular upgrades to TS 2.9 soon which should get rid of all of these issues. Regarding `detect-browser`, where do you see that `os` is required? I cannot find it in the source code at https://github.com/DamonOehlman/detect-browser/blob/master/index.js.

**Comment by crafter999** [comment]
Ignoring 'os' module will throw the following error and you check out this issue here for more information https://github.com/DamonOehlman/detect-browser/issues/54
`Module not found: Error: Can't resolve 'os' ......'node_modules/detect-browser'`

**Comment by edaniels** [comment]
Ah indeed. I left a comment for the author. If he doesn't get back within a week or so, I'll just fork, fix it, and submit a PR back. Sorry about that

**Comment by edaniels** [comment]
@crafter999 I just released 4.0.4 with the detect-browser change. Hopefully that helps the workaround slightly.

**Comment by edaniels** [comment]
And happy 4th of July. Just released 4.0.5 which should make us Angular 6 compatible. AFAICT, our code was actually breaking the type system and TS 2.9 allowed it somehow. My simple app seems to be working; let me know if yours does and I will close this issue.

**Comment by crafter999** [comment]
OK I've tested it and it works fine. Btw, I knew something was wrong at StitchClientErrorCode, StitchRequestErrorCode and RemoteInsertManyResult. I just couldn't messed up the whole project without clearly understanding every piece of code. Instead I made this dirty patch that was declaring some values as strings instead of numbers which was wrong but at least you got an idea where to fix things. I'm glad I was able to help in the end.
